spectec/spec/7-module.watsup

```diff
@@ -0,0 +1,236 @@
+
+def $alloc_module(s, module, externval*) = (s_res, m_res)
+  -- if module = MODULE import* func* global* table* mem* elem* data* start? export*
+  -- if m_init = { FUNC epsilon, GLOBAL epsilon, TABLE epsilon, MEM epsilon, ELEM epsilon, DATA epsilon, EXPORT epsilon }
+  -- if m_im = $alloc_import(m_init, import*, externval*)
+  -- if f = { LOCAL epsilon, MODULE m_im }
+  -- if (s_func, fa*) = $alloc_func((s; f), func*)
+  -- if (s_global, ga*) = $alloc_global((s_func; f), global*)
+  -- if (s_table, ta*) = $alloc_table((s_global; f), table*)
+  -- if (s_mem, ma*) = $alloc_mem((s_table; f), mem*)
+  -- if (s_elem, ea*) = $alloc_elem((s_mem; f), elem*)
+  -- if (s_data, da*) = $alloc_data((s_elem; f), data*)
+  -- if m_ex = m_im ++ { FUNC fa*, GLOBAL ga*, TABLE ta*, MEM ma*, ELEM ea*, DATA da*, EXPORT epsilon }
+  -- if xi* = $alloc_export(m_ex, export)*
+  -- if m_res = m_ex[.EXPORT = xi*]
+  -- if s_res = $replace_moduleinst(s_data, fa*, m_res)
+
+
+def $run_elem(state, elem*, idx) : state
+
+def $run_elem((s; f), epsilon, i) = s; f
+
+def $run_elem((s; f), elem elem'*, i) = s_res; f_res
+  -- if elem = ELEM reftype expr*
+  -- if $run_elem((s; f), elem'*, $(i + 1)) = s_res; f_res
+
+def $run_elem((s; f), elem elem'*, i) = s_res; f_res
+  -- if elem = ELEM reftype expr* elemmode
+  -- if n = |expr*|
+  -- if elemmode = TABLE x instr*
+  -- Step : s; f; instr* (CONST I32 0) (CONST I32 n) (TABLE.INIT x i) (ELEM.DROP i) ~> s_new; f_new; epsilon
+  -- if $run_elem((s_new; f_new), elem'*, $(i + 1)) = s_res; f_res
+
+def $run_elem((s; f), elem elem'*, i) = s_res; f_res
+  -- if elem = ELEM reftype expr* elemmode
+  -- if elemmode = DECLARE
+  -- Step : s; f; (ELEM.DROP i) ~> s_new; f_new; epsilon
+  -- if $run_elem((s_new; f_new), elem'*, $(i + 1)) = s_res; f_res
+
+
+def $run_data(state, data*, idx) : state
+
+def $run_data((s; f), epsilon, i) = s; f
+
+def $run_data((s; f), data data'*, i) = s_res; f_res
+  -- if data = DATA byte*
+  -- if $run_data((s; f), data'*, $(i + 1)) = s_res; f_res
+
+def $run_data((s; f), data data'*, i) = s_res; f_res
+  -- if data = DATA byte* datamode
+  -- if n = |byte*|
+  -- if datamode = MEMORY 0 instr*
+  -- Step : s; f; instr* (CONST I32 0) (CONST I32 n) (MEMORY.INIT i) (DATA.DROP i) ~> s_new; f_new; epsilon
+  -- if $run_data((s_new; f_new), data'*, $(i + 1)) = s_res; f_res
+
+
+;; Instantiation
+
+def $instantiation(store, module, externval*) : (store, moduleinst)
+
+def $instantiation(s, module, externval*) = (s_res, m)
+  -- if module = MODULE import* func* global* table* mem* elem* data* export*
+  -- if (s_alloc, m) = $alloc_module(s, module, externval*)
+  -- if f = { LOCAL epsilon, MODULE m }
+  -- if $run_elem((s_alloc; f), elem*, 0) = s_elem; f_elem
+  -- if $run_data((s_elem; f_elem), data*, 0) = s_res; f_res
+
+def $instantiation(s, module, externval*) = (s_res, m)
+  -- if module = MODULE import* func* global* table* mem* elem* data* start export*
```

Comment:
Can we avoid duplicating the definition? Why would the approach from the spec not work?